storage/src/snapshot_cache.rs

```diff
@@ -0,0 +1,268 @@
+
+    /// Attempts to insert a [`Snapshot`] entry into the cache. If the cache is full, the oldest
+    /// entry will be evicted to make space.
+    ///
+    /// [`Snapshot`]s must be inserted sequentially relative to their `jmt::Version`s and have
+    /// consecutive version numbers.
+    ///
+    /// ## Special case
+    ///
+    /// The first [`Snapshot`] inserted is allowed to have version `u64::MAX` to support the
+    /// pre-genesis case.
+    ///
+    /// ## Errors
+    ///
+    /// The method will return an error if the supplied `snapshot` has a version number that is:
+    ///
+    /// - stale i.e. older than the latest snapshots
+    ///
+    /// - skipping a version i.e. the difference between their version numbers is greater than 1
+    pub fn try_push(&mut self, snapshot: Snapshot) -> anyhow::Result<()> {
+        let latest_version = self.latest().version();
+        if latest_version.wrapping_add(1) != snapshot.version() {
+            return Err(anyhow!("snapshot_cache: trying to insert stale snapshots."));
+        }
+
+        if self.cache.len() >= self.max_size {
+            self.cache.pop_back();
+        }
+
+        self.cache.push_front(snapshot);
+        Ok(())
+    }
+
+    /// Inserts a [`Snapshot`] entry in the cache. If the cache is full, then the oldest entry will
+    /// be evicted to make space.
+    ///
+    /// [`Snapshot`]s must be inserted sequentially relative to their `jmt::Version`s and have
+    /// consecutive version numbers.
+    ///
+    /// ## Special case
+    ///
+    /// The first [`Snapshot`] inserted is allowed to have version `u64::MAX` to support the
+    /// pre-genesis case.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the supplied snapshot's version is:
+    ///
+    /// - stale i.e. older than the most recently inserted snapshot
+    ///
+    /// - skipping a version i.e. the difference between the version numbers is greater than 1.
+    ///
+    /// For a non-faillible version, see [`try_push`].
+    pub fn push(&mut self, snapshot: Snapshot) {
+        self.try_push(snapshot)
+            .expect("failed to push snapshot in cache");
+    }
+
+    /// Returns the latest inserted `Snapshot`.
+    pub fn latest(&self) -> Snapshot {
+        self.cache
+            .front()
+            .map(Clone::clone)
+            .expect("cache cannot be empty")
+    }
+
+    /// Returns the oldest snapshot in the cache.
+    fn oldest(&self) -> Snapshot {
+        self.cache.back().unwrap().clone()
+    }
+
+    /// Attempts to fetch a [`Snapshot`] with a matching `jmt::Version`, and returns `None` if none
+    /// was found.
+    pub fn get(&self, version: jmt::Version) -> Option<Snapshot> {
+        let latest_version = self.latest().version();
+
+        // Special case: the first ever snapshot inserted is allowed to have a "pre-genesis"
+        // `jmt::Version` version (`u64::MAX`). In this case, it is guaranteed to always be
+        // at the back of the queue.
```

Comment:
Do we need to special-case this?  Could we instead just compute the difference between the two versions with wrapping arithmetic, and use that as the offset?

Reply:
Then we could also get rid of the `oldest` method entirely, too.

Reply:
Good point, it's not a special case if we remove the `if version > latest_version` return early logic, cleaner indeed!